feat: introduce new GlobalHeader component (FE-5148) (#5269)
Sage/carbon: src/components/global-header/global-header.stories.mdx

import { useState } from "react";
import { Meta, Canvas, Story } from "@storybook/addon-docs";
import StyledSystemProps from "../../../.storybook/utils/styled-system-props";
import NavigationBar from "../navigation-bar";
import Box from "../box";
import VerticalDivider from "../vertical-divider";
import { Menu, MenuFullscreen, MenuItem, MenuSegmentTitle } from "../menu";
import GlobalHeader from "./global-header.component";
import useMediaQuery from "../../hooks/useMediaQuery";
import LinkTo from "@storybook/addon-links/react";
import carbonLogo from "../../../logo/carbon-logo.png";

<Meta
  title="Global Header"
  parameters={{
    info: { disable: true },
    chromatic: { disable: true },
    docs: {
      inlineStories: false,
      iframeHeight: 200,
    },
  }}
/>

# Global Header

Global Header is a wrapper component designed for creating site-wide navigation layouts.

## Contents

- [Quick Start](#quick-start)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import GlobalHeader from "carbon-react/lib/components/global-header";
```

## Examples

### Default

> Global Header shares similarities with the <LinkTo kind="Navigation Bar" story="default">Navigation Bar component</LinkTo>, but is permanently fixed to the top of the viewport and has a greater z-index than Navigation Bar.

Reviewer: ```suggestion
Global Header shares similarities with the <LinkTo kind="Navigation Bar" story="default">NavigationBar component</LinkTo>, but is permanently fixed to the top of the viewport and has a greater z-index than NavigationBar.
```

super picky I know but since this is referring to Carbon components I think it should use their actual component name, which of course does not have a space. I actually think the one in the link reads OK with the space and don't mind if that's left, but the sentence at the end I do think should have `NavigationBar` with no space.

<Canvas>
  <Story name="default">
    <GlobalHeader>Example content</GlobalHeader>
  </Story>
</Canvas>

### With logo

> A custom logo stored as a HTML element or a React component can be rendered by passing it via the `logo` prop. The logo is wrapped in a container that applies style rules for margins and child element height. It is possible to set a custom height for the logo element, but any value that exceeds the height of the Global Header component will be constrained to 40px.

Reviewer: ```suggestion
A custom logo in the form of an HTML element or a React component can be rendered by passing it via the `logo` prop. The logo is wrapped in a container that applies style rules for margins and child element height. It is possible to set a custom height for the logo element, but any value that exceeds the height of the GlobalHeader component will be constrained to 40px.
```

"stored as a" sounds strange to me - I'm not sure the best wording here and am not suggesting "in the form of" is great but I do slightly prefer it, feel free to come up with something better!

It should definitely be `an HTML element` rather than `a HTML element`. (You could simply say "a React element" which covers both HTML elements and React components, but I'm happy to be explicit because in practice the logo will usually be one of these 2 things, and not every developer will instantly know what "React element" means.)

Also, as above, when referencing `GlobalHeader` as a component then I prefer to give the component displayName, which doesn't have a space.

Reviewer 2: Pretty sure it's actually `a` when the following starts with and `h`

<Canvas>
  <Story name="with logo">
    {() => {
      const Logo = () => <img height={28} src={carbonLogo} alt="Carbon logo" />;
      return <GlobalHeader logo={<Logo />}>Example content</GlobalHeader>;
    }}
  </Story>
</Canvas>

### Basic menu

> In conjunction with the <LinkTo kind="Menu" story="default">Menu component</LinkTo>, the component can be used to create global navigation layouts.

Reviewer: ```suggestion
In conjunction with the <LinkTo kind="Menu" story="default">Menu component</LinkTo>, the `GlobalHeader` component can be used to create global navigation layouts.
```

Just "the component" is a bit ambiguous here, better I think to name the component you're talking about. (It could simply be `GlobalHeader` instead of "the `GlobalHeader` component".)

          <MenuItem href="#">Child Item 3</MenuItem>
        </MenuItem>,
        <MenuItem
          key="parent-menu-2"
          minWidth="145px"
          flex="0 0 auto"
          submenu="Parent Menu 2"
        >
          <MenuItem>Child Item</MenuItem>
        </MenuItem>,
      ];
      return (
        <GlobalHeader logo={<Logo />}>
          <Menu menuType="black" display="flex" flex="1">
            {fullscreenViewBreakPoint ? (
              <>
                <MenuItem
                  onClick={(ev) => {
                    ev.preventDefault();
                    setIsListViewOpen(true);
                  }}
                >
                  Admin
                </MenuItem>
                <MenuFullscreen
                  menuType="black"
                  isOpen={isListViewOpen}
                  onClose={() => setIsListViewOpen(false)}
                >
                  {menuItems}
                </MenuFullscreen>
              </>
            ) : (
              menuItems
            )}
          </Menu>
        </GlobalHeader>
      );
    }}
  </Story>
</Canvas>

### Global and local navigation bar layout

> The component can be used alongside the <LinkTo kind='Navigation Bar' story="default">Navigation Bar component</LinkTo> to create a two navigation bar layout, with the former being for site-wide navigation and the latter being for local navigation. Since Global Header is fixed to top of the viewport, the Navigation Bar must also have its `position` fixed, with a top `orientation` and `offset` of 40px.

Reviewer: ```suggestion
The component can be used alongside the <LinkTo kind='Navigation Bar' story="default">NavigationBar component</LinkTo> to create a two navigation bar layout, with the former being for site-wide navigation and the latter being for local navigation. Since GlobalHeader is fixed to top of the viewport, the NavigationBar must also have its `position` fixed, with a top `orientation` and `offset` of 40px.
```

Same comment as referenced a couple of times already.
